doc/dev/static_type_checking_cheat_sheet.md

```diff
@@ -0,0 +1,410 @@
+
+# No:
+T = TypeVar("T")
+
+class LROPoller(Generic[T]):
+    ...
+
+# Yes:
+PollingReturnType = TypeVar("PollingReturnType")
+
+class LROPoller(Generic[PollingReturnType]):
+    ...
+```
+
+- Follow naming conventions for covariant (`*_co`) and contravariant (`*_contra`) `TypeVar` parameters when setting variance for generic classes.
+
+```python
+from typing import TypeVar
+
+T_co = TypeVar("T_co", covariant=True)
+T_contra = TypeVar("T_contra", contravariant=True)
+```
+
+- Use the pre-defined TypeVar, `typing.AnyStr`, when your parameter or return type expects both `str` or `bytes`.
+- Do use `typing.ParamSpec` (a specialized TypeVar) to forward your function type hints from one callable to another callable (e.g. decorators). This preserves type hinting in the original function.
+
+```python
+from typing import TypeVar, Callable, ParamSpec
+
+
+T = TypeVar("T")
+P = ParamSpec("P")
+
+def validate_table(func: Callable[P, T]) -> Callable[P, T]:
+    def inner(*args: P.args, **kwargs: P.kwargs) -> T:
+        # validate table
+        return func(*args, **kwargs)
+    return inner
+
+
+@validate_table
+def create_table(table_name: str, **kwargs: Any) -> None:
+    ...
+```
+
+
+### Type Aliases
+
+- Do use type aliases to help make lengthy, complicated type hints more readable or help convey meaning to the reader.
+
+```python
+from typing import Union, Dict
+from azure.core.credentials import AzureKeyCredential, TokenCredential, AzureSasCredential
+
+CredentialTypes = Union[AzureKeyCredential, TokenCredential, AzureSasCredential, Dict[str, str]]  # PascalCase
+```
+
+- Do not use type aliases in docstrings. Specify the fully expanded type to satisfy rendered documentation.
+- Do use `typing.TypeAlias` if your type alias might be otherwise confused with a global assignment.
+
+### Literals
+
+- You can use `typing.Literal` when you want to restrict based on exact values.
+- Consider tagging an attribute with the `Literal` type to discriminate between models in a returned Union. This avoids the need to do `isinstance` checks on the return type.
+
+```python
+from typing import Union
+from typing_extensions import Literal
+
+
+# client library code
+class SelectionMark:
+
+    def __init__(self):
+        self.kind: Literal["selectionMark"] = "selectionMark"
+        self.state = "selected"
+
+
+class FormWord:
+
+    def __init__(self):
+        self.kind: Literal["word"] = "word"
+        self.word = "hello"
+
+
+class FormLine:
+
+    def __init__(self):
+        self.kind: Literal["line"] = "line"
+        self.line = "hello world"
+
+
+def get_element() -> Union[SelectionMark, FormWord, FormLine]:
+    ...
+
+
+# user code
+ele = get_element()
+
+if ele.kind == "selectionMark":
+    print(ele.state)
+elif ele.kind == "word":
+    print(ele.word)
+elif ele.kind == "line":
+    print(ele.line)
+```
+
+- If you have a `Literal` type that should not change or get re-assigned in the code, consider typing it as `typing.Final`.
```

Comment:
Maybe also mention the `@final` decorator for classes and methods? And `Final` variables can be used even without `Literal`.